mono/metadata/console-unix.c

```diff
@@ -216,50 +216,24 @@ tty_teardown (void)
 	setup_finished = FALSE;
 }
 
+static GENERATE_TRY_GET_CLASS_WITH_CACHE(system_console, "System", "Console")
+
 static void
 do_console_cancel_event (void)
 {
-	static MonoClassField *cancel_handler_field;
+	static MonoMethod *System_Console_DoConsoleCancelEventBackground_method = ((gpointer)-1);
 	MonoError error;
-	MonoDomain *domain = mono_domain_get ();
-	MonoClass *klass;
-	MonoDelegate *load_value;
-	MonoMethod *method;
-	MonoVTable *vtable;
-
-	/* FIXME: this should likely iterate all the domains, instead */
-	if (!domain->domain)
-		return;
 
-	klass = mono_class_try_load_from_name (mono_defaults.corlib, "System", "Console");
-	if (klass == NULL)
+	if (mono_defaults.console_class == NULL)
 		return;
 
-	if (cancel_handler_field == NULL) {
-		cancel_handler_field = mono_class_get_field_from_name (klass, "cancel_handler");
-		g_assert (cancel_handler_field);
-	}
-
-	vtable = mono_class_vtable_full (domain, klass, &error);
-	if (vtable == NULL || !is_ok (&error)) {
-		mono_error_cleanup (&error);
+	if (System_Console_DoConsoleCancelEventBackground_method == ((gpointer)-1))
+		System_Console_DoConsoleCancelEventBackground_method = mono_class_get_method_from_name (mono_defaults.console_class, "DoConsoleCancelEventInBackground", 0);
```

Comment:
This could be:
```c
if (System_Console_DoConsoleCancelEventBackground_method == ((gpointer)-1)) {
  MonoClass *system_console_class = mono_class_try_get_system_console_class ();
  if (system_console_class == NULL)
    return;
  System_Console_DoConsoleCancelEventBackground_method = mono_class_get_method_from_name (system_console_class, "DoConsoleCancelEventInBackground", 0);
}
if (System_Console_DoConsoleCancelEventBackground_method == NULL)
  return;
```